state/application.go

```diff
@@ -1648,14 +1648,21 @@ type SetCharmConfig struct {
 	RequireNoUnits bool
 }
 
-// SetCharm changes the charm for the application.
-func (a *Application) SetCharm(cfg SetCharmConfig) (err error) {
-	defer errors.DeferredAnnotatef(
-		&err, "cannot upgrade application %q to charm %q", a, cfg.Charm,
-	)
+func (a *Application) validateSetCharmConfig(cfg SetCharmConfig) error {
 	if cfg.Charm.Meta().Subordinate != a.doc.Subordinate {
 		return errors.Errorf("cannot change an application's subordinacy")
 	}
+	origin := cfg.CharmOrigin
+	if origin == nil {
+		return errors.NotValidf("nil charm origin")
+	}
+	if origin.Platform == nil {
+		return errors.BadRequestf("charm origin platform is nil")
+	}
+	if (origin.ID != "" && origin.Hash == "") || (origin.ID == "" && origin.Hash != "") {
+		return errors.BadRequestf("programming error, SetCharm, neither CharmOrigin ID nor Hash can be set before a charm is downloaded. See CharmHubRepository GetDownloadURL.")
+	}
+
```

Comment:
I wonder if this could be moved to a more general location?

Reply:
A `CharmOrigin.Validate` function perhaps? There is more validation we can do here, it would be good to have some maximally rigorous assertions somewhere